chromium_src/net/dns/dns_transaction.cc

```diff
@@ -0,0 +1,49 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "base/strings/string_util.h"
+#include "brave/components/unstoppable_domains/buildflags/buildflags.h"
+#include "net/dns/dns_config.h"
+#include "net/dns/dns_server_iterator.h"
+
+#if BUILDFLAG(UNSTOPPABLE_DOMAINS_ENABLED)
+#include "brave/components/unstoppable_domains/constants.h"
+#include "brave/components/unstoppable_domains/utils.h"
+#endif
+
+namespace {
+
+bool GetNextIndex(const std::string& hostname,
+                  const net::DnsConfig& config,
+                  net::DnsServerIterator* dns_server_iterator,
+                  size_t* doh_server_index) {
+#if BUILDFLAG(UNSTOPPABLE_DOMAINS_ENABLED)
+  // Skip unstoppable domains resolver for non-crypto domains.
+  if (unstoppable_domains::IsUnstoppableDomainsEnabled() &&
+      !base::EndsWith(hostname, unstoppable_domains::kCryptoDomain) &&
+      config.dns_over_https_servers[*doh_server_index].server_template ==
+          unstoppable_domains::kDoHResolver) {
+    // No next available index to attempt.
+    if (!dns_server_iterator->AttemptAvailable()) {
+      return false;
+    }
+
+    *doh_server_index = dns_server_iterator->GetNextAttemptIndex();
+  }
+#endif
+
+  return true;
+}
+
+}  // namespace
+
+#define BRAVE_MAKE_HTTP_ATTEMPT                                       \
+  if (!GetNextIndex(hostname_, session_.get()->config(),              \
```

Comment:
Just an idea - mb we could subclass `GetNextAttemptIndex()` to avoid this patch?

Reply:
@iefremov I considered that before, but there is a case where we cannot get an available next attempt index and need to return early in `MakeHTTPAttempt`.
This could happen when user is visiting non-crypto domain and they didn't specify their custom DoH provider, so there is only our UD resolver in the `config.dns_over_https_servers`, we don't have other servers to use in this case for non-crypto domain.